services/src/main/java/org/keycloak/services/resources/account/LinkedAccountsResource.java

```diff
@@ -0,0 +1,200 @@
+import javax.ws.rs.core.UriBuilder;
+import org.jboss.logging.Logger;
+import org.jboss.resteasy.spi.HttpRequest;
+import org.keycloak.common.util.Base64Url;
+import org.keycloak.credential.CredentialModel;
+import org.keycloak.events.Details;
+import org.keycloak.events.EventBuilder;
+import org.keycloak.events.EventType;
+import org.keycloak.models.AccountRoles;
+import org.keycloak.models.ClientModel;
+import org.keycloak.models.FederatedIdentityModel;
+import org.keycloak.models.IdentityProviderModel;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.UserModel;
+import org.keycloak.representations.account.AccountLinkUriRepresentation;
+import org.keycloak.representations.account.LinkedAccountsRepresentation;
+import org.keycloak.services.ErrorResponse;
+import org.keycloak.services.Urls;
+import org.keycloak.services.managers.Auth;
+import org.keycloak.services.messages.Messages;
+import org.keycloak.services.resources.Cors;
+import org.keycloak.services.validation.Validation;
+
+/**
+ * API for linking/unlinking social login accounts
+ *
+ * @author Stan Silvert
+ */
+public class LinkedAccountsResource {
+    private static final Logger logger = Logger.getLogger(LinkedAccountsResource.class);
+    
+    private final KeycloakSession session;
+    private final HttpRequest request;
+    private final ClientModel client;
+    private final EventBuilder event;
+    private final UserModel user;
+    private final RealmModel realm;
+    private final Auth auth;
+
+    public LinkedAccountsResource(KeycloakSession session, 
+                                  HttpRequest request, 
+                                  ClientModel client,
+                                  Auth auth, 
+                                  EventBuilder event, 
+                                  UserModel user) {
+        this.session = session;
+        this.request = request;
+        this.client = client;
+        this.auth = auth;
+        this.event = event;
+        this.user = user;
+        realm = session.getContext().getRealm();
+    }
+    
+    @GET
+    @Path("/")
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response linkedAccounts() {
+        LinkedAccountsBean linkedAccounts = new LinkedAccountsBean(this.session, this.realm, this.user);
+        LinkedAccountsRepresentation rep = new LinkedAccountsRepresentation();
+        rep.setHasAlternativeAuthentication(linkedAccounts.getHasAlternativeAuthentication());
+        rep.setLinkedAccounts(linkedAccounts.getLinkedAccounts());
+        return Cors.add(request, Response.ok(rep)).auth().allowedOrigins(auth.getToken()).build();
+    }
+    
+    @GET
+    @Path("/{providerId}")
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response buildLinkedAccountURI(@PathParam("providerId") String providerId, 
+                                     @QueryParam("redirectUri") String redirectUri) {
+        if (redirectUri == null) {
```

Comment:
We discovered that the client isn't able to check the nonce on redirect.   I think we considered that to be a weakness in the process today.

This endpoint is just for convenience and simplicity.  If we require the client to create the URI, then every client in every language has to write code for it, and it's not easy code to write.  See https://issues.jboss.org/browse/KEYCLOAK-7752 where a customer had problems doing this correctly.

The result is exactly the same whether the URI is generated on server or client.  This endpoint just makes it easier for everyone.

That being said, if we are definitely changing this stuff to AIA then we won't need the convenience endpoint going forward.

As you see from the JIRA, I have TypeScript code to take care of it, so if you still want me to get rid of the endpoint it won't cost me much extra development time.  But it does mean that the account console needs to download a crypto library that it wouldn't otherwise need.

Reply:
I definitively want to change it to an AIA, so would prefer if it's done on the server-side if it's not too much work.

Reply:
I'm feeling pressed for time getting the tech preview out.  I think I'll just leave it as-is for now but open a JIRA to convert this to using client-side before we go final.  Or better yet, do AIA before we go final.

Either way, this endpoint will only be used for tech preview.

Reply:
OK